### DISCUSSION TITLE
Repo reorganization

### DISCUSSION THREAD

**Discussion by Kralizek** [AutoFixture/AutoFixture#1190] [Ideas]
Hi @aivascu, 

would it be of any interest reorganizing the layout of the repository?

I'm looking specifically at the `Src` folder that contains the libraries and their test libraries. Usually, you see a `src` folder containing the libraries and a `tests` one containing the test projects.

A question would be what to do with all the solution files currently in the `Src` folder.

**Comment by aivascu** [comment]
@Kralizek the repo structure is something we have been discussing internally. There are some inconsistencies in the project/namespace structure. These inconsistencies range from, the suffix of the test projects (i.e. `UnitTest` vs `.UnitTest`) to whether some classes belong to the "Kernel" or not.

Having a `tests` folder is indeed common in other repos. In my experience the only benefit to this approach is the collection of test results for the CI, since you can write easier a glob expression (e.g. `tests/**/*.UnitTest`). 
Since I don't use an IDE like VS Code for .NET projects, the `.sln` gives me all the structure I need to organize the projects, but I imagine it would be an argument for those that do.

Now regarding the solution files in the `Src` folder. In the case we are going to move the test projects to a root level `tests` folder I  think the sensible solution would be to move the solution files, to the root of the repo so the `sln` would have consistent relative path to the `proj` files. The drawback would be that doing this would clutter the root folder of the repository.

In any case, I think the placement of the test projects and the solution files is more a matter of preference rather than utility. At the moment this does not create any urgent issues, so the reorganization of the test projects is low on the priority list.

**Comment by Kralizek** [comment]
I understand it's not an urgent issue but it's something I could work on my spare time.

An alternative approach would be to split by library rather than by src/tests.

Basically the structure could be

```
Src/
   AutoFixture.sln
   AutoFixture/
   AutoMoq/
      AutoFixture.AutoMoq.slnf
      AutoMoq/
      AutoMoqTests/
   ...

```

This would create a repository easier to navigate without cluttering the root with many `slnf` files.

I could put up a PR for you to look at if you want I don't think it will require much work.

**Comment by aivascu** [comment]
@Kralizek you could definitely try to propose the changes in a PR. However please consider that many current PR-s have been opened a while ago. Changing the folder structure may break them and the authors may not be around to rebase the PRs or fix conflicts.

My personal preference is to have a flat folder structure, with `sln`s int the root directory and a dedicated `tests` folder in case there are many projects. It helps when I write CI scripts or I want to quickly open a solution.
As for the structure within the solution, I organize the folders either by namespace or whatever structure I feel will make finding things easier. 
For this particular repository, I don't think any additional structure, within the solution, is necessary since the partial solution files already organize the projects into bounded contexts.

**Comment by jzabroski** [comment]
@aivascu It sounds like we should close stale PRs and open issues for people to go fix them if there is interest.  Such a master issue could be pinned.

I don't see PRs as true technical debt.  I used to, but after years of maintaining open source projects, I realize that people never come back to those PRs and I have little interest in fixing someone else's PR.  In this repo, perhaps the only exception is #928 which is something we shouldn't lose track of.

**Comment by aivascu** [comment]
@jzabroski I think that's a fair point. Thank you for sharing!
I will consider this while reviewing the backlog.
